modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java

```diff
@@ -122,6 +139,37 @@ public Expression run() {
         throw new IllegalArgumentException("expression engine does not know how to handle script context [" + context.name + "]");
     }
 
+    private static BucketAggregationScript.Factory newBucketAggregationScriptFactory(Expression expr) {
+        return () -> {
+            ReplaceableConstDoubleValues[] functionValuesArray =
+                new ReplaceableConstDoubleValues[expr.variables.length];
+            Map<String, ReplaceableConstDoubleValues> functionValuesMap = new HashMap<>();
+            for (int i = 0; i < expr.variables.length; ++i) {
+                functionValuesArray[i] = new ReplaceableConstDoubleValues();
+                functionValuesMap.put(expr.variables[i], functionValuesArray[i]);
+            }
+            return new BucketAggregationScript() {
+                @Override
+                public double execute(Map<String, Object> params) {
+                    params.forEach((name, value) -> {
```

Comment:
In expressions, we want these types of decisions made up front, not at execution time. Take a look at how params are passed to the factory in other script examples, and exposed as `getParams()` on the script class. We want to avoid costly things like hash lookups per document for something that won't change (eg the params that exist).

Reply:
@rjernst but these `params` aren't really `params` like in the other scripts. They aren't constant but change in a loop here https://github.com/elastic/elasticsearch/pull/32068/files#diff-15179b62bf7bf2f829791af279ea380aR90.

Previously those were part of a params field passed to the factory which made the script stateful and I guess led to this todo https://github.com/elastic/elasticsearch/pull/32068/files#diff-15179b62bf7bf2f829791af279ea380aL98 <= that I resolved here?

Reply:
Well, that isn't really resolving that comment. While only one script object is created with your change, the performance is the same because we are still doing the binding at execution time for each document. I would rather keep that original code with that comment, deprecate passing things in via params (by adding the parameters directly as arguments), and then fix the TODO in master.

Reply:
Or put differently :) 
If we don't move to passing the `params` as the `execute` argument, then we'll be forced to recreate the script over so long as we support passing variables to scripts via `params` won't we? => better move to more efficiently updating this map (and avoiding looping over it in expressions ideally) than adding the `params` field back and being forced to recreate the script?

Reply:
Then again, the correct solution will have `params` in this still -> will revert to using `params` as on the script tomorrow :)